File: ios/SanWissen/Content/SearchIndex.swift
import Foundation

/// Ein durchsuchbarer Treffer über alle Module hinweg — Entsprechung zu
/// `SearchItem` aus src/app/searchIndex.ts.
struct SearchItem: Identifiable, Hashable {
    let id: String
    let title: String
    let moduleId: String
    let moduleTitle: String
    let symbol: String
    let itemId: String
    let category: String
    /// Vorberechneter, kleingeschriebener Suchtext.
    let haystack: String
}

enum SearchIndex {

    static func build(from store: ContentStore) -> [SearchItem] {
        let categoryByModule = Dictionary(uniqueKeysWithValues: store.registry.modules.map { ($0.id, $0.category) })

        func make(moduleId: String, itemId: String, title: String, parts: [String?]) -> SearchItem {
            let info = store.module(id: moduleId)
            return SearchItem(
                id: "\(moduleId):\(itemId)",
                title: title,
                moduleId: moduleId,
                moduleTitle: info?.title ?? moduleId,
                symbol: info?.symbol ?? "square.grid.2x2",
                itemId: itemId,
                category: categoryByModule[moduleId] ?? "",
                haystack: parts.compactMap { $0 }.joined(separator: " ").lowercased())
        }

        var items: [SearchItem] = []

        for r in store.rhythms.rhythms {
            items.append(make(moduleId: "ekg", itemId: r.id, title: r.nameDe,
                              parts: [r.nameDe, r.nameEn] + r.keyFeatures))
        }

        for m in store.medikamente.medikamente {
            items.append(make(moduleId: "medikamente", itemId: m.id, title: m.name,
                              parts: [m.name, m.wirkstoff, m.arzneimittelgruppe, m.indikationen, m.wirkung]))
        }

        for (_, module) in store.topicModules {
            for t in module.topics {
                let facts = t.sections.flatMap { $0.items.map(\.text) }
                items.append(make(moduleId: module.moduleId, itemId: t.id, title: t.title,
                                  parts: [t.title, t.summary, t.category] + facts))
            }
        }

        for tool in store.tools {
            items.append(make(moduleId: "werkzeuge", itemId: tool.id, title: tool.title,
                              parts: [tool.title, tool.description, tool.category]))
        }

        for g in store.glossary {
            items.append(make(moduleId: "glossar", itemId: g.id, title: "\(g.abbr) — \(g.meaning)",
                              parts: [g.abbr, g.meaning, g.description]))
        }

        for c in store.checklists {
            items.append(make(moduleId: "checklisten", itemId: c.id, title: c.title,
                              parts: [c.title, c.description] + c.items.map(\.text)))
        }

        for card in store.cheatSheet {
            items.append(make(moduleId: "cheatsheet", itemId: card.id, title: card.title,
                              parts: [card.title] + card.points))
        }

        return items.sorted { $0.title.localizedStandardCompare($1.title) == .orderedAscending }
    }

    /// Sucht über alle Begriffe (UND-Verknüpfung), Treffer im Titel zuerst.
    static func search(_ query: String, in index: [SearchItem], limit: Int = 60) -> [SearchItem] {
        let terms = query.lowercased().split(separator: " ").map(String.init).filter { !$0.isEmpty }
        guard !terms.isEmpty else { return [] }

        let matches = index.filter { item in
            terms.allSatisfy { item.haystack.contains($0) }
        }
        return Array(matches.sorted { a, b in
            let aTitle = terms.allSatisfy { a.title.lowercased().contains($0) }
            let bTitle = terms.allSatisfy { b.title.lowercased().contains($0) }
            if aTitle != bTitle { return aTitle }
            return a.title.localizedStandardCompare(b.title) == .orderedAscending
        }.prefix(limit))
    }
}

File: src/app/searchIndex.ts
import modulesContent from '../../content/modules.json';
import { allTopicModules } from './content';
import { RHYTHMS } from '../modules/ekg/rhythms';
import { MEDIKAMENTE } from '../modules/medikamente/data';
import { TOOLS } from '../modules/werkzeuge/data';
import { GLOSSAR } from '../modules/glossar/data';
import { CHECKLISTEN } from '../modules/checklisten/data';
import { CHEATSHEET_CARDS } from '../modules/cheatsheet/data';

export interface SearchItem {
  key: string;
  title: string;
  moduleId: string;
  moduleTitle: string;
  icon: string;
  itemId: string;
  category: string;
  haystack: string;
}

/**
 * Durchsuchbarer Index über alle Module.
 *
 * Bis 1.1.0 stand hier für jedes Modul ein eigener Block, fünfzehn an der
 * Zahl, von denen sich zehn nur in Modul-ID und Feldnamen unterschieden.
 * Die zehn Themenmodule laufen jetzt über eine Schleife; übrig bleiben die
 * Quellen mit eigener Form.
 *
 * Titel, Icon und Kategorie je Modul kommen aus `content/modules.json`,
 * damit Suche und Seitenleiste nicht auseinanderlaufen. Die iOS-Suche
 * macht es genauso.
 */
const MODULE_INFO = new Map(modulesContent.modules.map((m) => [m.id, m]));

function make(moduleId: string, itemId: string, title: string, parts: (string | null | undefined)[]): SearchItem {
  const info = MODULE_INFO.get(moduleId);
  return {
    key: `${moduleId}:${itemId}`,
    title,
    moduleId,
    moduleTitle: info?.title ?? moduleId,
    icon: info?.icon ?? '',
    itemId,
    category: info?.category ?? '',
    haystack: parts.filter(Boolean).join(' ').toLowerCase(),
  };
}

const INDEX: SearchItem[] = [
  ...RHYTHMS.map((r) => make('ekg', r.id, r.nameDe, [r.nameDe, r.nameEn, ...r.keyFeatures])),

  ...MEDIKAMENTE.map((m) =>
    make('medikamente', m.id, m.name, [m.name, m.wirkstoff, m.arzneimittelgruppe, m.indikationen]),
  ),

  // Die zehn Module mit gemeinsamem Schema.
  ...allTopicModules().flatMap((mod) =>
    mod.topics.map((topic) =>
      make(mod.moduleId, topic.id, topic.title, [
        topic.title,
        topic.summary,
        ...topic.sections.flatMap((section) => section.items.map((item) => item.text)),
      ]),
    ),
  ),

  ...TOOLS.map((t) => make('werkzeuge', t.id, t.title, [t.title, t.description, t.category])),

  ...GLOSSAR.map((e) => make('glossar', e.id, e.abbr, [e.abbr, e.meaning, e.description])),

  ...CHECKLISTEN.map((c) =>
    make('checklisten', c.id, c.title, [c.title, c.description, ...c.items.map((i) => i.text)]),
  ),

  ...CHEATSHEET_CARDS.map((c) => make('cheatsheet', c.id, c.title, [c.title, ...c.points])),
];

export function searchAll(query: string, limit = 12): SearchItem[] {
  const q = query.trim().toLowerCase();
  if (!q) return [];
  return INDEX.filter((item) => item.haystack.includes(q)).slice(0, limit);
}
